### PR TITLE
[RLlib] CQL Fix

### PR DESCRIPTION
**Pull request by michaelzhiluo**
CQL Fix. Able to match OG paper results for 6 envs. Fairly confident it will work on other envs.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/cql/cql_torch_policy.py [hunk 1]
```diff
@@ -16,28 +16,35 @@
...
+from ray.rllib.utils.numpy import SMALL_NUMBER, MIN_LOG_NN_OUTPUT, \
+    MAX_LOG_NN_OUTPUT
 from ray.rllib.policy.policy import Policy
 from ray.rllib.policy.sample_batch import SampleBatch
 from ray.rllib.utils.framework import try_import_torch
 from ray.rllib.utils.typing import LocalOptimizer, TensorType, \
     TrainerConfigDict
 from ray.rllib.utils.torch_ops import apply_grad_clipping, \
     convert_to_torch_tensor
+from ray.rllib.utils.torch_ops import atanh
```
Comment (sven1977) [inline rllib/agents/cql/cql_torch_policy.py:28]:
```suggestion
    MAX_LOG_NN_OUTPUT
from ray.rllib.policy.policy import Policy
from ray.rllib.policy.sample_batch import SampleBatch
from ray.rllib.utils.framework import try_import_torch
from ray.rllib.utils.typing import LocalOptimizer, TensorType, \
    TrainerConfigDict
from ray.rllib.utils.torch_ops import apply_grad_clipping, atanh, \
    convert_to_torch_tensor
```

File: rllib/examples/env/d4rl_env.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import gym
+try:
+    import d4rl
+except ImportError:
+    d4rl = None
+"""
+8 Environments from D4RL Environment
```
Comment (sven1977) [inline rllib/examples/env/d4rl_env.py:7]:
```suggestion
8 Environments from D4RL Environment.
Use fully qualified class-path in your configs:
e.g. "env": "ray.rllib.examples.env.d4rl_env.halfcheetah_random".
```

**Review by sven1977 — APPROVED** [review]
Looks great! Thanks for digging in and providing the fixes and benchmarks! Waiting for tests to pass before merging ...

File: rllib/tuned_examples/sac/pendulum-sac.yaml [hunk 3]
```diff
@@ -8,7 +8,7 @@ pendulum-sac:
         timesteps_total: 10000
     config:
         # Works for both torch and tf.
-        framework: tf
+        framework: torch
```
Comment (sven1977) [inline rllib/tuned_examples/sac/pendulum-sac.yaml:11]:
```suggestion
        framework: tf
```

File: rllib/tuned_examples/sac/pendulum-sac.yaml [hunk 4]
```diff
@@ -32,7 +32,7 @@ pendulum-sac:
           critic_learning_rate: 0.0003
           entropy_learning_rate: 0.0003
         num_workers: 0
-        num_gpus: 0
+        num_gpus: 1
```
Comment (sven1977) [inline rllib/tuned_examples/sac/pendulum-sac.yaml:35]:
```suggestion
        num_gpus: 0
```

File: rllib/tuned_examples/cql/pendulum-cql.yaml [hunk 5]
```diff
@@ -0,0 +1,45 @@
+# Pendulum SAC can attain -150+ reward in 6-7k
```
Comment (sven1977) [inline rllib/tuned_examples/cql/pendulum-cql.yaml:1]: +100!

File: rllib/tuned_examples/cql/hopper-bc.yaml [hunk 6]
```diff
@@ -1,12 +1,20 @@
 hopper_bc:
-    env: Hopper-v2
+    env:
```
Comment (sven1977) [inline rllib/tuned_examples/cql/hopper-bc.yaml:2]: +100!

File: rllib/tuned_examples/cql/hopper-cql.yaml [hunk 7]
```diff
@@ -1,12 +1,20 @@
 hopper_cql:
-    env: Hopper-v2
+    env:
```
Comment (sven1977) [inline rllib/tuned_examples/cql/hopper-cql.yaml:2]: +100!

File: rllib/tuned_examples/cql/halfcheetah-cql.yaml [hunk 8]
```diff
@@ -1,12 +1,20 @@
 halfcheetah_cql:
-    env: HalfCheetah-v2
+    env:
```
Comment (sven1977) [inline rllib/tuned_examples/cql/halfcheetah-cql.yaml:2]: +100!

File: rllib/tuned_examples/cql/halfcheetah-bc.yaml [hunk 9]
```diff
@@ -1,12 +1,20 @@
 halfcheetah_bc:
-    env: HalfCheetah-v2
+    env:
```
Comment (sven1977) [inline rllib/tuned_examples/cql/halfcheetah-bc.yaml:2]: +100!

File: rllib/policy/policy_template.py [hunk 10]
```diff
@@ -170,6 +173,12 @@ def build_policy_class(
             Note: Only one of `make_model` or `make_model_and_action_dist`
             should be provided. If both are None, a default Model will be
             created.
+        compute_gradients_fn (Optional[Callable[
```
Comment (sven1977) [inline rllib/policy/policy_template.py:176]: Very cool, adding this flexibility.

File: rllib/agents/cql/cql.py [hunk 11]
```diff
@@ -127,13 +154,33 @@ def update_prio(item):
...
+            # and therefore force-set DONE=True to avoid this missing
+            # next-obs to cause learning problems.
+            if SampleBatch.NEXT_OBS not in batch:
+                obs = batch[SampleBatch.OBS]
+                batch[SampleBatch.NEXT_OBS] = \
+                    np.concatenate([obs[1:], np.zeros_like(obs[0:1])])
+                batch[SampleBatch.DONES][-1] = True
+            replay_buffer.add_batch(batch)
+    else:
+        print("Error here?, otherwise buffer would be empty!")
```
Comment (sven1977) [inline rllib/agents/cql/cql.py:183]:
```suggestion
        raise ValueError(
            "Unknown offline input! config['input'] must either be list of offline "
            "files (json) or a D4RL-specific InputReader specifier (e.g. "
            "'d4rl.hopper-medium-v0').")
```

File: rllib/tuned_examples/sac/pendulum-sac.yaml [hunk 12]
```diff
@@ -22,7 +22,7 @@ pendulum-sac:
         no_done_at_end: true
         n_step: 3
         rollout_fragment_length: 1
-        prioritized_replay: true
+        prioritized_replay: false
```
Comment (sven1977) [inline rllib/tuned_examples/sac/pendulum-sac.yaml:25]:
```suggestion
        prioritized_replay: true
```

**Comment by michaelzhiluo** [comment]
@sven1977 Travis passed! Should be good to go

**Comment by sven1977** [comment]
Looks great! Learning Pendulum-v0 now as well from a simple json file (no D4RL!).
Just waiting for tests to pass. Thanks again @michaelzhiluo ! :)

**Pull request merged**